feature: makes poly-commit compatible with snarkOS

kobigurk opened the pull request:
Ports to the snarkos traits, and uses the new serialization derives

kobigurk committed:
feature: makes poly-commit compatible with snarkOS

gakonst commented on src/data_structures.rs:
```diff
@@ -65,7 +66,7 @@ pub trait PCRandomness: Clone {
-pub trait PCProof: Clone + algebra_core::ToBytes {
+pub trait PCProof: Clone + ToBytes {
     /// Size in bytes
     fn size_in_bytes(&self) -> usize;
```
Since we got `CanonicalSerialize` now, this method could be replaced with the `serialized_size` method's implementation. Same comment applies to `size_in_bytes` above in `PcCommitment`

gakonst commented on src/data_structures.rs:
```diff
@@ -39,7 +40,7 @@ pub trait PCVerifierKey: Clone + core::fmt::Debug {
-pub trait PCCommitment: Clone + algebra_core::ToBytes {
+pub trait PCCommitment: Clone + ToBytes {
     /// Outputs a non-hiding commitment to the zero polynomial.
     fn empty() -> Self;
```
Can't we use the Zero trait instead of `empty`? (same for all other traits which have this function)

gakonst commented on src/data_structures.rs:
```diff
@@ -186,9 +187,12 @@ impl<C: PCCommitment> LabeledCommitment<C> {
     }
 }
 
-impl<C: PCCommitment> algebra_core::ToBytes for LabeledCommitment<C> {
```
Should we instead implement `CanonicalSerialize`?

gakonst commented on src/ipa_pc/data_structures.rs:
```diff
@@ -105,13 +106,16 @@ impl<G: AffineCurve> PCCommitment for Commitment<G> {
 
     fn size_in_bytes(&self) -> usize {
-        algebra_core::to_bytes![G::zero()].unwrap().len() / 2
+        to_bytes![G::zero()].unwrap().len() / 2
```
I am advocating for the CanonicalSerialize trait, because for example in this case we'd be able to just use `G::SERIALIZED_SIZE` (or `G::UNCOMPRESSED_SIZE` if you'd want an uncompressed size). 

From a performance perspective, this call right now does multiple allocations (get zero element + convert to bytes) vs getting the trait's constant

gakonst commented on src/ipa_pc/data_structures.rs:
```diff
@@ -105,13 +106,16 @@ impl<G: AffineCurve> PCCommitment for Commitment<G> {
     }
 }
 
 impl<G: AffineCurve> ToBytes for Commitment<G> {
```
Here, you could `#[derive(CanonicalSerialize, CanonicalDeserialize)]` on the struct instead of having to manually impl it.

Same applies for all follow-up impls of the `size_in_bytes` and the `write` functions inside their respective traits

gakonst commented on src/ipa_pc/data_structures.rs:
```diff
@@ -192,13 +196,16 @@ pub struct Proof<G: AffineCurve> {
 impl<G: AffineCurve> PCProof for Proof<G> {
     fn size_in_bytes(&self) -> usize {
-        algebra_core::to_bytes![self].unwrap().len()
+        to_bytes![self].unwrap().len()
```
ditto

gakonst commented on src/ipa_pc/data_structures.rs:
```diff
@@ -192,13 +196,16 @@ pub struct Proof<G: AffineCurve> {
 
 impl<G: AffineCurve> ToBytes for Proof<G> {
     #[inline]
-    fn write<W: algebra_core::io::Write>(&self, mut writer: W) -> algebra_core::io::Result<()> {
```
ditto

gakonst commented on src/kzg10/data_structures.rs:
```diff
@@ -91,7 +94,10 @@ pub struct Commitment<E: PairingEngine>(
 
 impl<E: PairingEngine> ToBytes for Commitment<E> {
     #[inline]
-    fn write<W: algebra_core::io::Write>(&self, writer: W) -> algebra_core::io::Result<()> {
```
ditto

gakonst commented on src/kzg10/data_structures.rs:
```diff
@@ -107,7 +113,7 @@ impl<E: PairingEngine> PCCommitment for Commitment<E> {
 
     fn size_in_bytes(&self) -> usize {
-        algebra_core::to_bytes![E::G1Affine::zero()].unwrap().len() / 2
+        to_bytes![E::G1Affine::zero()].unwrap().len() / 2
```
ditto

gakonst commented on src/kzg10/data_structures.rs:
```diff
@@ -221,17 +227,20 @@ pub struct Proof<E: PairingEngine> {
     fn size_in_bytes(&self) -> usize {
         let hiding_size = if self.random_v.is_some() {
-            algebra_core::to_bytes![E::Fr::zero()].unwrap().len()
+            to_bytes![E::Fr::zero()].unwrap().len()
```
```suggestion
            E::Fr::UNCOMPRESSED_SIZE
```

gakonst replied:
This is a bit weird. The hiding size uses the uncompressed Fr size, while the G1Affine uses the compressed one (since it divides by 2?)?

gakonst commented on src/kzg10/data_structures.rs:
```diff
@@ -221,17 +227,20 @@ pub struct Proof<E: PairingEngine> {
             0
         };
-        algebra_core::to_bytes![E::G1Affine::zero()].unwrap().len() / 2 + hiding_size
+        to_bytes![E::G1Affine::zero()].unwrap().len() / 2 + hiding_size
```
```suggestion
        E::G1Affine::SERIALIZED_SIZE + hiding_size
```

gakonst commented on src/kzg10/data_structures.rs:
```diff
@@ -221,17 +227,20 @@ pub struct Proof<E: PairingEngine> {
     }
 }
 
 impl<E: PairingEngine> ToBytes for Proof<E> {
```
`derive[(CanonicalSerialize, CanonicalDeserialize)]` instead of manual impl

gakonst commented on src/kzg10/mod.rs:
```diff
@@ -100,17 +104,17 @@ impl<E: PairingEngine> KZG10<E> {
         let h = h.into_affine();
-        let prepared_h = h.into();
-        let prepared_beta_h = beta_h.into();
+        let prepared_h = h.prepare();
```
Do you have a list of these compatibility changes somewhere? Can see it being useful that we have it published somewhere

gakonst commented on src/lib.rs:
```diff
@@ -12,12 +12,14 @@
 
 #[macro_use]
 extern crate derivative;
```
nit: Extern crate syntax is no longer required, you can do `use derivative::Derivative` instead

gakonst commented on src/marlin_pc/data_structures.rs:
```diff
@@ -144,7 +145,10 @@ pub struct Commitment<E: PairingEngine> {
 
 impl<E: PairingEngine> ToBytes for Commitment<E> {
     #[inline]
-    fn write<W: algebra_core::io::Write>(&self, mut writer: W) -> algebra_core::io::Result<()> {
```
derive as well?

gakonst reviewed (approved):
LGTM, a big list of nits from my end because I feel we only stand to gain by deriving the 2 serialization traits I ported over from Zexe, but functionally this is good

gakonst committed:
chore: remove ipa and sonic poly-commits

gakonst committed:
chore: fmt

gakonst committed:
feat: add macro for ToBytes/FromBytes

This delegates the implementation to the derived CanonicalSerialize and Deserialize traits

gakonst committed:
feat: add Canonical(de)Serialize trait bounds to all poly traits

gakonst committed:
feat(kzg10): derive Canonical(De)Serialize and delegate to ToBytes

gakonst committed:
feat(marlin-pc): derive Canonical(De)Serialize and delegate to ToBytes

gakonst committed:
fix: remove size_in_bytes functions

These are not necessary, since we can use the CanonicalSerialize trait's serialized_size functions

gakonst committed:
chore: reorder cargo imports

gakonst committed:
fix(labeled-commitment): serialize the LabeledCommitment without the labels

gakonst committed:
fix: s/delegate/impl_bytes

gakonst committed:
Merge pull request #3 from AleoHQ/gakonst/snarkos-serializers

Implement missing SnarkOS Serializers

gakonst commented:
I've resolved the nits, this should be good to merge once we get CI working.

howardwu committed:
Adds snarkOS CI integration (#5)

* Update ci.yml

* Adds rustfmt, precommit hook, and cargo fmts repo

* Update toolchain to nightly

* Comment out no_std ci check

* Update name for secret
